pkg/factory/factory.go

```diff
@@ -0,0 +1,174 @@
+		return resources.NewUioResource()
+	default:
+		return resources.NewGenericResource()
+	}
+}
+
+// GetSelector returns an instance of DeviceSelector using selector attribute string and its associated values
+func (rf *resourceFactory) GetSelector(attr string, values []string) (types.DeviceSelector, error) {
+	switch attr {
+	case "vendors":
+		return resources.NewVendorSelector(values), nil
+	case "devices":
+		return resources.NewDeviceSelector(values), nil
+	case "drivers":
+		return resources.NewDriverSelector(values), nil
+	case "pfNames":
+		return resources.NewPfNameSelector(values), nil
+	case "linkTypes":
+		return resources.NewLinkTypeSelector(values), nil
+	case "ddpProfiles":
+		return resources.NewDdpSelector(values), nil
+	default:
+		return nil, fmt.Errorf("GetSelector(): invalid attribute %s", attr)
+	}
+}
+
+// GetResourcePool returns an instance of resourcePool
+func (rf *resourceFactory) GetResourcePool(rc *types.ResourceConfig, filteredDevice []types.PciDevice) (types.ResourcePool, error) {
+
+	devicePool := make(map[string]types.PciDevice, 0)
+	apiDevices := make(map[string]*pluginapi.Device)
+	for _, dev := range filteredDevice {
+		pciAddr := dev.GetPciAddr()
+		devicePool[pciAddr] = dev
+		apiDevices[pciAddr] = dev.GetAPIDevice()
+		glog.Infof("device added: [pciAddr: %s, vendor: %s, device: %s, driver: %s]",
+			dev.GetPciAddr(),
+			dev.GetVendor(),
+			dev.GetDeviceCode(),
+			dev.GetDriver())
+	}
+
+	var rPool types.ResourcePool
+	var err error
+	switch rc.DeviceType {
+	case types.NetDeviceType:
+		if len(filteredDevice) > 0 {
+			if _, ok := filteredDevice[0].(types.PciNetDevice); ok {
+				rPool = netdevice.NewNetResourcePool(rc, apiDevices, devicePool)
+			} else {
+				rPool = nil
+				err = fmt.Errorf("invalid device list for NetDeviceType")
+			}
+		}
+	case types.AcceleratorType:
+		if len(filteredDevice) > 0 {
+			if _, ok := filteredDevice[0].(types.AccelDevice); ok {
+				rPool = accelerator.NewAccelResourcePool(rc, apiDevices, devicePool)
+			} else {
+				rPool = nil
+				err = fmt.Errorf("invalid device list for AcceleratorType")
+			}
+		}
+	default:
+		err = fmt.Errorf("cannot create resourcePool: invalid device type %s", rc.DeviceType)
+	}
+	return rPool, err
+}
+
+func (rf *resourceFactory) GetRdmaSpec(pciAddrs string) types.RdmaSpec {
+	return netdevice.NewRdmaSpec(pciAddrs)
+}
+
+// GetDeviceProvider returns an instance of DeviceProvider based on DeviceType
+func (rf *resourceFactory) GetDeviceProvider(dt types.DeviceType) types.DeviceProvider {
+	switch dt {
+	case types.NetDeviceType:
+		return netdevice.NewNetDeviceProvider(rf)
+	case types.AcceleratorType:
+		return accelerator.NewAccelDeviceProvider(rf)
+	default:
+		return nil
+	}
+}
+
+// GetDeviceFilter unmarshal the "selector" values from ResourceConfig and returns an instance of DeviceSelector based on
+// DeviceType in the ResourceConfig
+func (rf *resourceFactory) GetDeviceFilter(rc *types.ResourceConfig) (interface{}, error) {
```

Comment:
So providing an unsupported selector will fail silently, that's OK ?
you could add some logic to enforce that.

Reply:
For invalid deviceType it will not fail silently. As you can see, errors are propagated from here, manager will not to save this config for creating resource pool and will be logged.

Reply:
so it silently ignores unsupported selectors of a certain *valid* device type.
i dont understand your comment

Reply:
for valid device type, this method will try to unmarshal selector configuration. Any error occurred during this will also be returned and subsequently be logged by the manager. I don't see anything silently ignored, am I missing something?

Reply:
So im most likely missing something, what will happen if i provide the following resource config ?
```
 {
            "resourceName": "intel_fpga",
            "deviceType": "accelerator",
            "selectors": {
                    "vendors": ["8086"],
                    "devices": ["0d90"],
                    "pfNames": ["enp0s0f0","enp2s2f1"]
            }
        }
```

will it fail with something resembling "pfNames selector is invalid for deviceType accelerator" ?

Reply:
Ok, with this particular example it will not fail because json unmarshal will correctly fill in vendors and devices fields in the AccelDeviceSelectors struct but will ignore pfNames as there is no matching field for it there. But if we had a type mismatch for example int in "vendor" will result in error and will be captured during unmarshalling. Also, if there were some mandatory fields for AccelDeviceSelectors (without 'omitempty' json tag) and that was not given in the selectors that would be captured as well during unmarshalling.

Reply:
Indeed, its provided by unmarshall :)

so going back to my initial comment: 
> So providing an unsupported selector will fail silently, that's OK ?

clarifying unsupported as unsupported by the device type.

you could first unmarshal it into a map[string]interface{} check the keys then unmarshal to the final struct

Reply:
I still don't get the "silently failing". Giving arbitrary fields in json will always be ignored if that's what you call silently failing.

Reply:
edited, my comment above, i hope its clearer.

Reply:
> Indeed, its provided by unmarshall :)
> 
> Clarifying my initial comment:
> 
> So providing an unsupported selector by the device type will ignore it silently, that's OK ?

I think this is how generally json objects are handle - CNI would be a perfect example. Only cares about the fields that a plugin actually need.

> 
> you could first unmarshal it into a map[string]interface{} check the keys then unmarshal to the final struct if silently failing is not desired

This would be a little overkill and we still will be left with same situation. Say for example, we first unmarshall the selectors into the map[string]interface{} and which will give us something like this: "vendors": interface{} and it's still unknown what is the actual value given for the vendors.  So, if there was some `int` value given instead of `[]string` we still could not capture this error here.

Reply:
i see, OK.